tensorflow/core/grappler/optimizers/auto_mixed_precision_lists.h

```diff
@@ -0,0 +1,300 @@
+  static std::set<string> GrayList() {
+    if (IsPseudoFastMath()) {
+      return std::set<string>{};
+    }
+    string to_add, to_remove;
+    ReadStringFromEnvVar("TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_GRAYLIST_ADD",
+                         "", &to_add);
+    ReadStringFromEnvVar(
+        "TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_GRAYLIST_REMOVE", "",
+        &to_remove);
+
+    auto list = std::set<string>{
+        "Add",
+        "AddN",
+        "AddV2",
+        "AvgPool",
+        "AvgPool3D",
+        "AvgPool3DGrad",
+        "AvgPoolGrad",
+        "BiasAdd",
+        "BiasAddGrad",
+        "BiasAddV1",
+        "Elu",
+        "EluGrad",
+        "Erf",
+        "Erfc",
+        "FloorDiv",
+        "FusedBatchNormV2",
+        "FusedBatchNormGradV2",
+        "Inv",
+        "LeakyRelu",
+        "LeakyReluGrad",
+        "Mul",
+        "Prod",
+        "RealDiv",
+        "Reciprocal",
+        "Sigmoid",
+        "SigmoidGrad",
+        "Softplus",
+        "SoftplusGrad",
+        "Sqrt",
+        "Sub",
+        "Tanh",
+        "TanhGrad",
+    };
+    UpdateList(&list, to_add, to_remove);
+    return list;
+  }
+
+  // Returns the set of ops that are considered numerically-dangerous (i.e.,
+  // unsafe for execution in fp16) and whose effects may also be observed in
+  // downstream nodes (e.g., in Exp -> Add, the Add is unsafe due to the Exp).
+  static std::set<string> BlackList() {
+    if (IsPseudoFastMath()) {
+      return std::set<string>{};
+    }
+    string to_add, to_remove;
+    ReadStringFromEnvVar("TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_BLACKLIST_ADD",
+                         "", &to_add);
+    ReadStringFromEnvVar(
+        "TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_BLACKLIST_REMOVE", "",
+        &to_remove);
+
+    auto list = std::set<string>{
+        "Exp",
+        "Expm1",
+        "L2Loss",
+        "Log",
+        "Log1p",
+        "LogSoftmax",
+        "Mean",
+        "Pow",
+        "SaveV2",
+        "Softmax",
+        "SoftmaxCrossEntropyWithLogits",
+        "SparseSoftmaxCrossEntropyWithLogits",
+        "Sum",
+    };
+    UpdateList(&list, to_add, to_remove);
+    return list;
+  }
+
+  // Returns the set of ops that do not have numerically-significant effects
+  // (i.e., they are always considered safe for execution in fp16 precision).
+  static std::set<string> ClearList() {
+    if (IsPseudoFastMath()) {
+      return std::set<string>{};
+    }
+    // Note: if a stateful op (such as StackPopV2) is added to the clearlist,
+    // you must also modify the AutoMixedPrecisionImpl class to call
+    // AddDataStructureOpsToMap() with that op
+    string to_add, to_remove;
+    ReadStringFromEnvVar("TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_CLEARLIST_ADD",
+                         "", &to_add);
+    ReadStringFromEnvVar(
+        "TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_CLEARLIST_REMOVE", "",
+        &to_remove);
+
+    auto list = std::set<string>{
```

Comment:
We usually use this pattern for large static string sets: (from grappler/op_types.cc)

static const gtl::FlatSet<string>* const kNonForwardingOps = CHECK_NOTNULL(
      (new gtl::FlatSet<string>{"ArgMax",
                                "ArgMin",
                                "AudioSpectrogram",
                                "BatchMatMul",
...

Same everywhere. std::set is very slow and can fragment memory.

Reply:
Sounds good, thanks.  I replaced the std::set instances with gtl::FlatSet, let me know if I should change it.